.github/workflows/build.yml

```
@@ -0,0 +1,64 @@
+name: Build struct2tensor wheels
+
+on:
+  pull_request:
+  workflow_dispatch:
+  release:
+    types: [published]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Build package
+        id: build-package
+        uses: ./.github/reusable-build
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: List and check wheels
+        shell: bash
+        run: |
+          # Pin transitive dependency on pkginfo until https://github.com/pypa/twine/issues/1070 is
+          # fixed
+          pip install twine pkginfo>=1.10.0
+          ${{ matrix.ls || 'ls -lh' }} dist/
+          twine check dist/*
+
+  upload_to_pypi:
+    name: Upload to PyPI
+    runs-on: ubuntu-latest
+    if: (github.event_name == 'release' && startsWith(github.ref, 'refs/tags')) || (github.event_name == 'workflow_dispatch')
+    needs: [build]
+    environment:
+      name: pypi
+      url: https://pypi.org/p/struct2tensor/
+    permissions:
+      id-token: write
```

```suggestion
      id-token: write
      contents: read
      attestations: write
```

- `contents: read`: better to be explicit rather than implicit about this on the runner that's supposed to publish the wheels
- `attestations: write`: for the GHA Attestations feature mentioned below (please feel free to remove this if it's not needed).